File: src/TemporalProlog/Scenario.hs
-- |
-- Module      : TemporalProlog.Scenario
-- Description : Portable bounded model-checking scenario format
--
-- A scenario names a Temporal Prolog program, fixes a finite exploration
-- horizon, supplies ground assertions at selected worlds, and declares
-- safety invariants as forbidden atom patterns.  The deliberately small,
-- line-oriented format is also implemented by the Rust engine.
module TemporalProlog.Scenario
  ( Invariant(..)
  , ChoiceAlternative(..)
  , ChoiceGroup(..)
  , Scenario(..)
  , parseScenario
  ) where

import Control.Monad (foldM, unless, when)
import Data.Char (isAlphaNum, isSpace)
import qualified Data.Map.Strict as Map
import Data.Map.Strict (Map)
import qualified Data.Set as Set
import Data.Set (Set)
import Text.Megaparsec (errorBundlePretty)
import Text.Read (readMaybe)

import TemporalProlog.Parser (parseAtom)
import TemporalProlog.Syntax (Atom, isGroundAtom)

data Invariant = Invariant
  { invariantName      :: String
  , invariantForbidden :: Atom
  } deriving (Eq, Ord, Show)

data ChoiceAlternative
  = ChoiceAtom Atom
  | ChoiceNone
  deriving (Eq, Ord, Show)

-- | Exactly one alternative in each named group is supplied at its world.
-- Multiple groups at the same world form a Cartesian product.
data ChoiceGroup = ChoiceGroup
  { choiceGroupName         :: String
  , choiceGroupAlternatives :: [ChoiceAlternative]
  } deriving (Eq, Ord, Show)

data Scenario = Scenario
  { scenarioName       :: String
  , scenarioProgram    :: FilePath
  , scenarioSteps      :: Int
  , scenarioAssertions :: Map Int [Atom]
  , scenarioChoices    :: Map Int [ChoiceGroup]
  , scenarioInvariants :: [Invariant]
  } deriving (Eq, Show)

data PartialScenario = PartialScenario
  { partialName           :: Maybe String
  , partialProgram        :: Maybe FilePath
  , partialSteps          :: Maybe Int
  , partialAssertions     :: Map Int [Atom]
  , partialChoices        :: Map Int [ChoiceGroup]
  , partialInvariants     :: [Invariant]
  , partialInvariantNames :: Set String
  }

emptyPartial :: PartialScenario
emptyPartial = PartialScenario
  { partialName = Nothing
  , partialProgram = Nothing
  , partialSteps = Nothing
  , partialAssertions = Map.empty
  , partialChoices = Map.empty
  , partialInvariants = []
  , partialInvariantNames = Set.empty
  }

-- | Parse a portable @.tpmc@ scenario.  Directives occupy one line each:
--
-- @
-- name example
-- program example.tpl
-- steps 4
-- assert 0 request(1)
-- choose 1 participant vote_yes(tx, participant)
-- choose 1 participant vote_no(tx, participant)
-- invariant mutual_exclusion forbids violation(mutual_exclusion)
-- @
parseScenario :: FilePath -> String -> Either String Scenario
parseScenario sourceName source = do
  partial <- foldM parseNumberedLine emptyPartial (zip [1 :: Int ..] (lines source))
  name <- requireField "name" (partialName partial)
  program <- requireField "program" (partialProgram partial)
  steps <- requireField "steps" (partialSteps partial)
  when (steps <= 0) $ Left "steps must be a positive integer"
  mapM_ (validateInputStep "assertion" steps) (Map.keys (partialAssertions partial))
  mapM_ (validateInputStep "choice" steps) (Map.keys (partialChoices partial))
  mapM_ validateChoiceGroups (Map.toList (partialChoices partial))
  Right Scenario
    { scenarioName = name
    , scenarioProgram = program
    , scenarioSteps = steps
    , scenarioAssertions = partialAssertions partial
    , scenarioChoices = partialChoices partial
    , scenarioInvariants = partialInvariants partial
    }
  where
    parseNumberedLine partial (lineNumber, original) =
      let line = trim (takeWhile (/= '%') original)
      in if null line
           then Right partial
           else firstLineError lineNumber (parseDirective lineNumber line partial)

    firstLineError lineNumber = either
      (Left . ((sourceName ++ ":" ++ show lineNumber ++ ": ") ++))
      Right

    validateInputStep kind steps step
      | step < 0 = Left $ kind ++ " steps must be non-negative"
      | step >= steps = Left $ kind ++ " step " ++ show step
          ++ " is outside the " ++ show steps ++ "-world horizon"
      | otherwise = Right ()

    validateChoiceGroups (step, groups) = mapM_ (validateChoiceGroup step) groups
    validateChoiceGroup step group
      | length (choiceGroupAlternatives group) < 2 = Left $
          "choice group '" ++ choiceGroupName group ++ "' at step " ++ show step
          ++ " must have at least two alternatives"
      | otherwise = Right ()

requireField :: String -> Maybe a -> Either String a
requireField field = maybe (Left $ "missing required '" ++ field ++ "' directive") Right

parseDirective :: Int -> String -> PartialScenario -> Either String PartialScenario
parseDirective lineNumber line partial =
  let (directive, rest) = splitWord line
  in case directive of
    "name" -> do
      value <- exactlyOne "name" rest
      unless (validName value) $ Left "scenario names may contain only letters, digits, '_' and '-'"
      ensureUnset "name" (partialName partial)
      Right partial { partialName = Just value }
    "program" -> do
      let value = trim rest
      when (null value) $ Left "program requires a relative or absolute path"
      ensureUnset "program" (partialProgram partial)
      Right partial { partialProgram = Just value }
    "steps" -> do
      value <- exactlyOne "steps" rest
      count <- maybe (Left "steps requires a positive integer") Right (readMaybe value)
      ensureUnset "steps" (partialSteps partial)
      Right partial { partialSteps = Just count }
    "assert" -> do
      let (stepText, atomText) = splitWord rest
      when (null stepText || null atomText) $ Left "assert requires STEP and ATOM"
      step <- maybe (Left "assert step must be an integer") Right (readMaybe stepText)
      atom <- parseScenarioAtom lineNumber atomText
      unless (isGroundAtom atom) $ Left "scheduled assertions must be ground"
      Right partial
        { partialAssertions = Map.insertWith (flip (++)) step [atom]
            (partialAssertions partial)
        }
    "choose" -> do
      let (stepText, afterStep) = splitWord rest
          (groupName, alternativeText) = splitWord afterStep
      when (null stepText || null groupName || null alternativeText) $
        Left "choose requires STEP GROUP and ATOM or 'none'"
      step <- maybe (Left "choose step must be an integer") Right (readMaybe stepText)
      unless (validName groupName) $
        Left "choice group names may contain only letters, digits, '_' and '-'"
      alternative <- if alternativeText == "none"
        then Right ChoiceNone
        else do
          atom <- parseScenarioAtom lineNumber alternativeText
          unless (isGroundAtom atom) $ Left "choice alternatives must be ground"
          Right (ChoiceAtom atom)
      choices <- addChoice step groupName alternative (partialChoices partial)
      Right partial { partialChoices = choices }
    "invariant" -> do
      let (name, afterName) = splitWord rest
          (keyword, atomText) = splitWord afterName
      when (null name || keyword /= "forbids" || null atomText) $
        Left "invariant requires NAME forbids ATOM"
      unless (validName name) $ Left "invariant names may contain only letters, digits, '_' and '-'"
      when (name `Set.member` partialInvariantNames partial) $
        Left $ "duplicate invariant name '" ++ name ++ "'"
      atom <- parseScenarioAtom lineNumber atomText
      let invariant = Invariant name atom
      Right partial
        { partialInvariants = partialInvariants partial ++ [invariant]
        , partialInvariantNames = Set.insert name (partialInvariantNames partial)
        }
    _ -> Left $ "unknown directive '" ++ directive ++ "'"
  where
    parseScenarioAtom n text = case parseAtom ("<scenario line " ++ show n ++ ">") text of
      Left err -> Left $ "invalid atom:\n" ++ errorBundlePretty err
      Right atom -> Right atom

addChoice
  :: Int
  -> String
  -> ChoiceAlternative
  -> Map Int [ChoiceGroup]
  -> Either String (Map Int [ChoiceGroup])
addChoice step groupName alternative choices = do
  groups <- case Map.lookup step choices of
    Nothing -> Right [ChoiceGroup groupName [alternative]]
    Just existing -> update existing
  Right (Map.insert step groups choices)
  where
    update [] = Right [ChoiceGroup groupName [alternative]]
    update (group:rest)
      | choiceGroupName group /= groupName = (group :) <$> update rest
      | alternative `elem` choiceGroupAlternatives group = Left $
          "duplicate alternative in choice group '" ++ groupName ++ "'"
      | otherwise = Right
          (group { choiceGroupAlternatives = choiceGroupAlternatives group ++ [alternative] } : rest)

ensureUnset :: String -> Maybe a -> Either String ()
ensureUnset _ Nothing = Right ()
ensureUnset field (Just _) = Left $ "duplicate '" ++ field ++ "' directive"

exactlyOne :: String -> String -> Either String String
exactlyOne directive rest = case words rest of
  [value] -> Right value
  _ -> Left $ directive ++ " requires exactly one value"

validName :: String -> Bool
validName name = not (null name) && all (\c -> isAlphaNum c || c == '_' || c == '-') name

splitWord :: String -> (String, String)
splitWord input =
  let stripped = dropWhile isSpace input
      (word, rest) = break isSpace stripped
  in (word, trim rest)

trim :: String -> String
trim = reverse . dropWhile isSpace . reverse . dropWhile isSpace

File: rust/src/scenario.rs
use crate::{parse_atom, Atom};
use std::collections::{BTreeMap, BTreeSet};

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct Invariant {
    pub name: String,
    pub forbidden: Atom,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub enum ChoiceAlternative {
    Atom(Atom),
    None,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct ChoiceGroup {
    pub name: String,
    pub alternatives: Vec<ChoiceAlternative>,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct Scenario {
    pub name: String,
    pub program: String,
    pub steps: usize,
    pub assertions: BTreeMap<usize, Vec<Atom>>,
    pub choices: BTreeMap<usize, Vec<ChoiceGroup>>,
    pub invariants: Vec<Invariant>,
}

#[derive(Default)]
struct PartialScenario {
    name: Option<String>,
    program: Option<String>,
    steps: Option<usize>,
    assertions: BTreeMap<usize, Vec<Atom>>,
    choices: BTreeMap<usize, Vec<ChoiceGroup>>,
    invariants: Vec<Invariant>,
    invariant_names: BTreeSet<String>,
}

/// Parse the portable line-oriented `.tpmc` scenario format shared with the
/// Haskell implementation.
pub fn parse_scenario(source_name: &str, source: &str) -> Result<Scenario, String> {
    let mut partial = PartialScenario::default();
    for (index, original) in source.lines().enumerate() {
        let line_number = index + 1;
        let line = original.split('%').next().unwrap_or_default().trim();
        if line.is_empty() {
            continue;
        }
        parse_directive(line_number, line, &mut partial)
            .map_err(|error| format!("{source_name}:{line_number}: {error}"))?;
    }

    let name = partial
        .name
        .ok_or_else(|| "missing required 'name' directive".to_string())?;
    let program = partial
        .program
        .ok_or_else(|| "missing required 'program' directive".to_string())?;
    let steps = partial
        .steps
        .ok_or_else(|| "missing required 'steps' directive".to_string())?;
    if steps == 0 {
        return Err("steps must be a positive integer".into());
    }
    for step in partial.assertions.keys() {
        if *step >= steps {
            return Err(format!(
                "assertion step {step} is outside the {steps}-world horizon"
            ));
        }
    }
    for (step, groups) in &partial.choices {
        if *step >= steps {
            return Err(format!(
                "choice step {step} is outside the {steps}-world horizon"
            ));
        }
        for group in groups {
            if group.alternatives.len() < 2 {
                return Err(format!(
                    "choice group '{}' at step {step} must have at least two alternatives",
                    group.name
                ));
            }
        }
    }

    Ok(Scenario {
        name,
        program,
        steps,
        assertions: partial.assertions,
        choices: partial.choices,
        invariants: partial.invariants,
    })
}

fn parse_directive(
    line_number: usize,
    line: &str,
    partial: &mut PartialScenario,
) -> Result<(), String> {
    let (directive, rest) = split_word(line);
    match directive {
        "name" => {
            let value = exactly_one("name", rest)?;
            if !valid_name(value) {
                return Err("scenario names may contain only letters, digits, '_' and '-'".into());
            }
            if partial.name.replace(value.to_string()).is_some() {
                return Err("duplicate 'name' directive".into());
            }
        }
        "program" => {
            if rest.is_empty() {
                return Err("program requires a relative or absolute path".into());
            }
            if partial.program.replace(rest.to_string()).is_some() {
                return Err("duplicate 'program' directive".into());
            }
        }
        "steps" => {
            let value = exactly_one("steps", rest)?;
            let steps = value
                .parse::<usize>()
                .map_err(|_| "steps requires a positive integer")?;
            if partial.steps.replace(steps).is_some() {
                return Err("duplicate 'steps' directive".into());
            }
        }
        "assert" => {
            let (step, atom_source) = split_word(rest);
            if step.is_empty() || atom_source.is_empty() {
                return Err("assert requires STEP and ATOM".into());
            }
            let step = step
                .parse::<usize>()
                .map_err(|_| "assert step must be an integer")?;
            let atom = parse_atom(atom_source)
                .map_err(|error| format!("invalid atom on scenario line {line_number}: {error}"))?;
            if !atom.ground() {
                return Err("scheduled assertions must be ground".into());
            }
            partial.assertions.entry(step).or_default().push(atom);
        }
        "choose" => {
            let (step, after_step) = split_word(rest);
            let (group_name, alternative_source) = split_word(after_step);
            if step.is_empty() || group_name.is_empty() || alternative_source.is_empty() {
                return Err("choose requires STEP GROUP and ATOM or 'none'".into());
            }
            let step = step
                .parse::<usize>()
                .map_err(|_| "choose step must be an integer")?;
            if !valid_name(group_name) {
                return Err(
                    "choice group names may contain only letters, digits, '_' and '-'".into(),
                );
            }
            let alternative = if alternative_source == "none" {
                ChoiceAlternative::None
            } else {
                let atom = parse_atom(alternative_source).map_err(|error| {
                    format!("invalid atom on scenario line {line_number}: {error}")
                })?;
                if !atom.ground() {
                    return Err("choice alternatives must be ground".into());
                }
                ChoiceAlternative::Atom(atom)
            };
            add_choice(&mut partial.choices, step, group_name, alternative)?;
        }
        "invariant" => {
            let (name, after_name) = split_word(rest);
            let (keyword, atom_source) = split_word(after_name);
            if name.is_empty() || keyword != "forbids" || atom_source.is_empty() {
                return Err("invariant requires NAME forbids ATOM".into());
            }
            if !valid_name(name) {
                return Err("invariant names may contain only letters, digits, '_' and '-'".into());
            }
            if !partial.invariant_names.insert(name.to_string()) {
                return Err(format!("duplicate invariant name '{name}'"));
            }
            let forbidden = parse_atom(atom_source)
                .map_err(|error| format!("invalid atom on scenario line {line_number}: {error}"))?;
            partial.invariants.push(Invariant {
                name: name.to_string(),
                forbidden,
            });
        }
        _ => return Err(format!("unknown directive '{directive}'")),
    }
    Ok(())
}

fn add_choice(
    choices: &mut BTreeMap<usize, Vec<ChoiceGroup>>,
    step: usize,
    group_name: &str,
    alternative: ChoiceAlternative,
) -> Result<(), String> {
    let groups = choices.entry(step).or_default();
    if let Some(group) = groups.iter_mut().find(|group| group.name == group_name) {
        if group.alternatives.contains(&alternative) {
            return Err(format!(
                "duplicate alternative in choice group '{group_name}'"
            ));
        }
        group.alternatives.push(alternative);
    } else {
        groups.push(ChoiceGroup {
            name: group_name.to_string(),
            alternatives: vec![alternative],
        });
    }
    Ok(())
}

fn exactly_one<'a>(directive: &str, rest: &'a str) -> Result<&'a str, String> {
    let mut values = rest.split_whitespace();
    let value = values
        .next()
        .ok_or_else(|| format!("{directive} requires exactly one value"))?;
    if values.next().is_some() {
        return Err(format!("{directive} requires exactly one value"));
    }
    Ok(value)
}

fn valid_name(name: &str) -> bool {
    !name.is_empty()
        && name
            .chars()
            .all(|character| character.is_alphanumeric() || matches!(character, '_' | '-'))
}

fn split_word(input: &str) -> (&str, &str) {
    let input = input.trim_start();
    input
        .find(char::is_whitespace)
        .map_or((input, ""), |index| {
            let (word, rest) = input.split_at(index);
            (word, rest.trim())
        })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_portable_scenario() {
        let scenario = parse_scenario(
            "test.tpmc",
            "name demo\nprogram demo.tpl\nsteps 2\nassert 0 request(1)\n\
             invariant safe forbids violation(X)\n",
        )
        .unwrap();
        assert_eq!(scenario.name, "demo");
        assert_eq!(scenario.steps, 2);
        assert_eq!(scenario.assertions[&0].len(), 1);
        assert_eq!(scenario.invariants.len(), 1);
    }

    #[test]
    fn parses_independent_choice_groups() {
        let scenario = parse_scenario(
            "test.tpmc",
            "name demo\nprogram demo.tpl\nsteps 2\n\
             choose 1 p1 yes(p1)\nchoose 1 p1 no(p1)\n\
             choose 1 p2 yes(p2)\nchoose 1 p2 none\n",
        )
        .unwrap();
        assert_eq!(scenario.choices[&1].len(), 2);
        assert_eq!(scenario.choices[&1][0].alternatives.len(), 2);
    }

    #[test]
    fn rejects_out_of_horizon_assertion() {
        let error = parse_scenario(
            "test.tpmc",
            "name demo\nprogram demo.tpl\nsteps 1\nassert 1 request\n",
        )
        .unwrap_err();
        assert!(error.contains("outside"));
    }
}
